src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,164 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import classNames from 'classnames';
+import { Todo } from '../../types/Todo';
+import { FC, useEffect, useRef, useState } from 'react';
+
+interface Props {
+  todo: Todo;
+  isLoading?: boolean;
+  handleDelete?: () => Promise<void>; // 🔧 змінено тип функції
+  isDeleting: boolean;
+  onToggle?: () => Promise<void>;
+  onRename?: (title: string) => Promise<void>;
+}
+
+export const TodoItem: FC<Props> = ({
+  todo,
+  isLoading,
+  handleDelete,
+  isDeleting,
+  onToggle,
+  onRename,
+}: Props) => {
+  const [isToggleLoading, setIsToggleLoading] = useState(false);
+  const [isFormActive, setIsFormActive] = useState(false);
+  const [isLoadingChanges, setIsLoadingChanges] = useState(false);
+  const [enterTriggeredSave, setEnterTriggeredSave] = useState(false);
+  const [title, setTitle] = useState(todo.title);
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  useEffect(() => {
+    if (isFormActive && inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [isFormActive]);
+
+  const handleToggleTodo = async () => {
+    setIsToggleLoading(true);
+    if (onToggle) {
+      await onToggle();
+    }
+
+    setIsToggleLoading(false);
+  };
+
+  const focusInput = () => {
+    setTimeout(() => inputRef.current?.focus(), 0);
+  };
+
+  const saveChanges = async () => {
+    setIsLoadingChanges(true);
+
+    const trimmedTitle = title.trim();
+    const originalTitle = todo.title.trim();
+
+    try {
+      if (!trimmedTitle) {
+        try {
+          if (handleDelete) {
+            await handleDelete();
+          }
+
+          setIsFormActive(false);
+
+          return;
+        } catch {
+          focusInput();
+
+          return;
+        }
+      }
+
+      if (trimmedTitle !== originalTitle) {
+        await onRename?.(trimmedTitle);
+      }
+
+      setIsFormActive(false);
+    } catch {
+      focusInput();
+    } finally {
+      setIsLoadingChanges(false);
+      setEnterTriggeredSave(false);
+    }
+  };
```

Comment:
Try to think how we can optimaize this code for reading and better understanding. Also you need use one of way in work with promise async/await or than/catch notation